spec/unit/middleware/timeout_spec.rb

```diff
@@ -0,0 +1,177 @@
+require 'spec_helper'
+require 'qless/job'
+require 'qless/middleware/timeout'
+
+module Qless
+  module Middleware
+    ::Rspec.describe Timeout do
+      class JobClass
+        attr_accessor :perform
+
+        def around_perform_call_counter
+          @around_perform_call_counter ||= 0
+        end
+
+        def around_perform(job)
+          @around_perform_call_counter = (@around_perform_call_counter || 0) + 1
+        end
+      end
+
+      let(:kernel_class) { class_double(Kernel) }
+
+      let(:job) {
+        instance_double(Qless::Job)
+      }
+
+      def make_worker(timeout_class, timeout_seconds, kernel_class)
+        Class.new(JobClass) do
+          include Qless::Middleware::Timeout.new(timeout_class: timeout_class,
+              kernel_class: kernel_class) { timeout_seconds }
+        end.new
+      end
+
+      before {
+        allow(job).to receive(:reconnect_to_redis)
+        allow(job).to receive(:klass_name)
+        allow(job).to receive(:fail)
+        allow(kernel_class).to receive(:exit!)
+      }
+
+      class TriggeredTimeout
+        def self.timeout(timeout_seconds)
+          raise ::Timeout::Error.new("triggered at #{timeout_seconds}s")
+        end
+      end
+
+      class InactiveTimeout
+        def self.timeout(timeout_seconds)
+          yield
+        end
+      end
+
+      context 'when timeout specified as nil (to block timeout processing for specific jobs)' do
+        it 'allows nil' do
+          worker = make_worker(TriggeredTimeout, nil, kernel_class)
+
+          expect {
+            worker.around_perform job
+          }.not_to raise_error
+        end
+
+        it 'invokes job' do
+          worker = make_worker(TriggeredTimeout, nil, kernel_class)
+          worker = worker
+
+          worker.around_perform job
+
+          expect(worker.around_perform_call_counter).to eq(1)
+        end
+      end
+
+      context 'when timeout is not nil' do
+        it 'aborts with a clear error when given a non-numeric timeout' do
+          worker = make_worker(TriggeredTimeout, "123", kernel_class)
+
+          expect {
+            worker.around_perform job
+          }.to raise_error(Qless::InvalidTimeoutError)
+        end
+
+        it 'aborts with a clear error when given a non-positive timeout' do
```

Comment:
Is it worth making another spec for the `0` case?  This doc string states that `0` should have the same behavior as the `-1` below but there is no example demonstrating that.  (Not a big deal to me either way but perhaps worth doing).